### PR TITLE
Add table aws_emr_cluster. Closes #141

### LINKED ISSUE #141: Add table aws_emr_cluster
**References**
https://pkg.go.dev/github.com/citysir/aws-sdk-go/service/emr

### PR DESCRIPTION
**Pull request by khushboo9024**
```No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/aws_emr_cluster []

PRETEST: tests/aws_emr_cluster

TEST: tests/aws_emr_cluster
Running terraform
data.aws_region.alternate: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.aws_partition.current: Refreshing state...
data.aws_region.primary: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_key_pair.deployer: Creating...
aws_iam_role.iam_emr_profile_role: Creating...
aws_iam_role.iam_emr_service_role: Creating...
aws_vpc.my_vpc: Creating...
aws_key_pair.deployer: Creation complete after 2s [id=deployer-key]
aws_iam_role.iam_emr_service_role: Creation complete after 4s [id=turbottest35598_1]
aws_iam_role.iam_emr_profile_role: Creation complete after 4s [id=turbottest35598_2]
aws_iam_instance_profile.emr_profile: Creating...
aws_iam_role_policy.iam_emr_service_policy: Creating...
aws_iam_role_policy.iam_emr_profile_policy: Creating...
aws_iam_role_policy.iam_emr_profile_policy: Creation complete after 2s [id=turbottest35598_2:turbottest35598_2]
aws_iam_role_policy.iam_emr_service_policy: Creation complete after 3s [id=turbottest35598_1:turbottest35598_1]
aws_iam_instance_profile.emr_profile: Creation complete after 3s [id=var.resource_name]
aws_vpc.my_vpc: Still creating... [10s elapsed]
aws_vpc.my_vpc: Creation complete after 12s [id=vpc-0ea8ec306dd1f99f6]
aws_internet_gateway.gw: Creating...
aws_subnet.my_subnet: Creating...
aws_subnet.my_subnet: Creation complete after 3s [id=subnet-09e66bbc81d0d2bbd]
aws_security_group.emr_master: Creating...
aws_internet_gateway.gw: Creation complete after 5s [id=igw-09d39a187759455d3]
aws_route_table.test_route_table: Creating...
aws_route_table.test_route_table: Creation complete after 4s [id=rtb-0ac9cba38ecae133d]
aws_main_route_table_association.test_association: Creating...
aws_security_group.emr_master: Creation complete after 8s [id=sg-062a17541936828e3]
aws_emr_cluster.named_test_resource: Creating...
aws_main_route_table_association.test_association: Creation complete after 2s [id=rtbassoc-045dccc4ff72f6c0d]
aws_emr_cluster.named_test_resource: Still creating... [10s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [20s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [30s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [40s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [50s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [1m0s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [1m10s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [1m20s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [1m30s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [1m40s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [1m50s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [2m0s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [2m10s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [2m20s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [2m30s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [2m40s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [2m50s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [3m0s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [3m10s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [3m20s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [3m30s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [3m40s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [3m50s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [4m0s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [4m10s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [4m20s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [4m30s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [4m40s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [4m50s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [5m0s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [5m10s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [5m20s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [5m30s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [5m40s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [5m50s elapsed]
aws_emr_cluster.named_test_resource: Still creating... [6m0s elapsed]
aws_emr_cluster.named_test_resource: Creation complete after 6m0s [id=j-1KFDE6CCQIEDG]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Warning: Quoted references are deprecated

  on variables.tf line 201, in resource "aws_security_group" "emr_master":
 201:   depends_on = ["aws_subnet.my_subnet"]

In this context, references are expected literally rather than in quotes.
Terraform 0.11 and earlier required quotes, but quoted references are now
deprecated and will be removed in a future version of Terraform. Remove the
quotes surrounding this reference to silence this warning.

(and 2 more similar warnings elsewhere)


Apply complete! Resources: 13 added, 0 changed, 0 destroyed.

Outputs:

account_id = 013122550996
region_name = us-east-1
resource_aka = arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-1KFDE6CCQIEDG
resource_id = j-1KFDE6CCQIEDG
resource_name = turbottest35598
transit_gateway_id = j-1KFDE6CCQIEDG
vpc_id = vpc-0ea8ec306dd1f99f6

Running SQL query: test-get-query.sql
[
  {
    "auto_terminate": false,
    "cluster_arn": "arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-1KFDE6CCQIEDG",
    "ebs_root_volume_size": "<null>",
    "id": "j-1KFDE6CCQIEDG",
    "name": "turbottest35598",
    "tags": {
      "name": "turbottest35598"
    }
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "account_id": "013122550996",
    "akas": [
      "arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-1KFDE6CCQIEDG"
    ],
    "auto_terminate": false,
    "cluster_arn": "arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-1KFDE6CCQIEDG",
    "id": "j-1KFDE6CCQIEDG",
    "name": "turbottest35598",
    "region": "us-east-1",
    "tags": {
      "name": "turbottest35598"
    }
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "auto_terminate": false,
    "cluster_arn": "arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-1KFDE6CCQIEDG",
    "id": "j-1KFDE6CCQIEDG",
    "name": "turbottest35598"
  }
]
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-1KFDE6CCQIEDG"
    ],
    "tags": {
      "name": "turbottest35598"
    },
    "title": "turbottest35598"
  }
]
✔ PASSED

POSTTEST: tests/aws_emr_cluster

TEARDOWN: tests/aws_emr_cluster

SUMMARY:

1/1 passed.



```

```
### Basic info

+----------------+------------------------------------------------------------------------+------+----------------+------------------------+------+
| id             | cluster_arn                                                            | name | auto_terminate | state                  | tags |
+----------------+------------------------------------------------------------------------+------+----------------+------------------------+------+
| j-1MBZDQQ2Z5JH | arn:aws:elasticmapreduce:us-east-2:013122550996:cluster/j-1MBZDQQ2Z5JH | test | false          | TERMINATED_WITH_ERRORS | {}   |
+----------------+------------------------------------------------------------------------+------+----------------+------------------------+------+



### List of EMR clusters whose auto terminate is disabled

+-----------------+-------------------------------------------------------------------------+----------------+
| name            | cluster_arn                                                             | auto_terminate |
+-----------------+-------------------------------------------------------------------------+----------------+
| My cluster-1    | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-2T9N1TR7AJFDV | false          |
| turbottest14260 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-1NJH4IH6D920Q | false          |
| test            | arn:aws:elasticmapreduce:us-east-2:013122550996:cluster/j-1MBZDQQ2Z5JH  | false          |
| turbottest77471 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-3T1GFDKFTU4MI | false          |
| My cluster-2    | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-U43QEEANA25X  | false          |
| turbottest94111 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-2FREY17M6HLXG | false          |
| turbottest31444 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-1269PQY4A72R5 | false          |
| My cluster      | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-335UES9W40Z8W | false          |
+-----------------+-------------------------------------------------------------------------+----------------+


### List of EMR clusters which got terminated due to some error

+-----------------+--------------+------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------
| id              | name         | state                  | state_change_reason                                                                                                                             
+-----------------+--------------+------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------
| j-2T9N1TR7AJFDV | My cluster-1 | TERMINATED_WITH_ERRORS | The VPC/subnet configuration was invalid: No route to any external sources detected in Route Table for Subnet: subnet-06eddb9d6773ab2e6 for VPC:
| j-U43QEEANA25X  | My cluster-2 | TERMINATED_WITH_ERRORS | The VPC/subnet configuration was invalid: No route to any external sources detected in Route Table for Subnet: subnet-091d40f38b6a3230f for VPC:
| j-1MBZDQQ2Z5JH  | test         | TERMINATED_WITH_ERRORS | The VPC/subnet configuration was invalid: Subnet subnet-ab18bae7 has a route in blackhole state. This means that the route's target is unavailab
+-----------------+--------------+------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------


### List of application names and its versions installed on EMR clusters

+------+------------------------------------------------------------------------+------------------+---------------------+
| name | cluster_arn                                                            | application_name | application_version |
+------+------------------------------------------------------------------------+------------------+---------------------+
| test | arn:aws:elasticmapreduce:us-east-2:013122550996:cluster/j-1MBZDQQ2Z5JH | Hadoop           | 2.10.1              |
| test | arn:aws:elasticmapreduce:us-east-2:013122550996:cluster/j-1MBZDQQ2Z5JH | Hive             | 2.3.7               |
| test | arn:aws:elasticmapreduce:us-east-2:013122550996:cluster/j-1MBZDQQ2Z5JH | Hue              | 4.8.0               |
| test | arn:aws:elasticmapreduce:us-east-2:013122550996:cluster/j-1MBZDQQ2Z5JH | Mahout           | 0.13.0              |
| test | arn:aws:elasticmapreduce:us-east-2:013122550996:cluster/j-1MBZDQQ2Z5JH | Pig              | 0.17.0              |
| test | arn:aws:elasticmapreduce:us-east-2:013122550996:cluster/j-1MBZDQQ2Z5JH | Tez              | 0.9.2               |
+------+------------------------------------------------------------------------+------------------+---------------------+




### List clusters for which logging is disabled

```sql
select
  name,
  cluster_arn,
  log_uri
from
  aws_emr_cluster
where
  log_uri is null
  ```
  ```
  +-----------------+-------------------------------------------------------------------------+---------+
| name            | cluster_arn                                                             | log_uri |
+-----------------+-------------------------------------------------------------------------+---------+
| test            | arn:aws:elasticmapreduce:us-east-2:013122550996:cluster/j-1MBZDQQ2Z5JH  | <null>  |
| turbottest63433 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-3FGNSJ7AP807P | <null>  |
| turbottest32112 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-3DI0B0SJNT9YO | <null>  |
| turbottest38626 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-39CW4P2P3XLP7 | <null>  |
| turbottest38495 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-19EJ0IJXSI5TG | <null>  |
| turbottest60306 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-3O3F02NIO6GR9 | <null>  |
| turbottest20572 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-2OZLU4MOMAI6Y | <null>  |
| turbottest86064 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-LLE6SJUGTIHV  | <null>  |
| turbottest97237 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-VJ3KWJUB2IZS  | <null>  |
| turbottest94111 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-2FREY17M6HLXG | <null>  |
| turbottest14260 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-1NJH4IH6D920Q | <null>  |
```
```
### List clusters for which logging is enabled but log encryption is disabled

```sql
select
  name,
  cluster_arn,
  log_uri,
  log_encryption_kms_key_id
from
  aws_emr_cluster
where
  log_uri is not null and log_encryption_kms_key_id is null;
  ```+--------------+-------------------------------------------------------------------------+---------------------------------------------------------+---------------------------+
| name         | cluster_arn                                                             | log_uri                                                 | log_encryption_kms_key_id |
+--------------+-------------------------------------------------------------------------+---------------------------------------------------------+---------------------------+
| My cluster   | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-X051MBNSQFCC  | s3n://aws-logs-013122550996-us-east-1/elasticmapreduce/ | <null>                    |
| My cluster   | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-2HMRDX853VYZL | s3n://aws-logs-013122550996-us-east-1/elasticmapreduce/ | <null>                    |
| My cluster-1 | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-2T9N1TR7AJFDV | s3n://aws-logs-013122550996-us-east-1/elasticmapreduce/ | <null>                    |
| My cluster   | arn:aws:elasticmapreduce:us-east-1:013122550996:cluster/j-335UES9W40Z8W | s3n://aws-logs-013122550996-us-east-1/elasticmapreduce/ | <null>                    |
+--------------+-------------------------------------------------------------------------+---------------------------------------------------------+---------------------------+

```
```
```

### CODE REVIEW

File: aws/table_aws_emr_cluster.go [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+			{
+				Name:        "cluster_arn",
+				Description: "The Amazon Resource Name of the cluster.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "name",
+				Description: "The name of the cluster.",
+				Type:        proto.ColumnType_STRING,
+			},
```
Comment (Subhajit97) [inline aws/table_aws_emr_cluster.go:44]: Use `name` as first column

File: aws/table_aws_emr_cluster.go [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+			{
+				Name:        "Normalized_instance_hours",
```
Comment (Subhajit97) [inline aws/table_aws_emr_cluster.go:46]:
```suggestion
				Name:        "normalized_instance_hours",
```

Comment (Subhajit97) [inline aws/table_aws_emr_cluster.go:46]:
```suggestion
				Name:        "normalized_instance_hours",
```

File: aws/table_aws_emr_cluster.go [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+				Description: "The friendly name to identify the DB Cluster.",
```
Comment (Subhajit97) [inline aws/table_aws_emr_cluster.go:47]: Update the description

File: aws/table_aws_emr_cluster.go [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+			},
+			{
+				Name:        "kerberos_attributes",
+				Description: "Attributes for Kerberos configuration when Kerberos authentication is enabled using a security configuration.",
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     getEmrCluster,
+			},
+			{
+				Name:        "tags_src",
+				Description: "A list of tags associated with a cluster..",
```
Comment (Subhajit97) [inline aws/table_aws_emr_cluster.go:202]:
```suggestion
				Description: "A list of tags associated with a cluster.",
```

File: aws/table_aws_emr_cluster.go [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+//// TRANSFORM FUNCTIONS
+
+func getEmrClusterTurbotTags(ctx context.Context, d *transform.TransformData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("emrClusterTagToTurbotTags")
+	clusterTags := d.HydrateItem.(*emr.Cluster)
+
+	if clusterTags == nil {
+		return nil, nil
+	}
+	// Mapping the resource tags inside turbotTags
```
Comment (Subhajit97) [inline aws/table_aws_emr_cluster.go:312]:
```suggestion
	
	// Mapping the resource tags inside turbotTags
```

File: aws/table_aws_emr_cluster.go [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+			},
+			{
+				Name:        "custom_ami_id",
+				Description: "Available only in Amazon EMR version 5.7.0 and later. The ID of a custom Amazon EBS-backed Linux AMI if the cluster uses a custom AMI.",
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getEmrCluster,
+			},
+			{
+				Name:        "ebs_root_volume_size",
+				Description: "The size, in GiB, of the Amazon EBS root device volume of the Linux AMI that is used for each EC2 instance. Available in Amazon EMR version 4.x and later.",
```
Comment (Subhajit97) [inline aws/table_aws_emr_cluster.go:77]:
```suggestion
				Description: "The size of the Amazon EBS root device volume of the Linux AMI that is used for each EC2 instance, in GiB. Available in Amazon EMR version 4.x and later.",
```

File: aws/table_aws_emr_cluster.go [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+				Hydrate:     getEmrCluster,
+			},
+			{
+				Name:        "tags_src",
+				Description: "A list of tags associated with a cluster..",
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     getEmrCluster,
+				Transform:   transform.FromField("Tags"),
+			},
+
```
Comment (Subhajit97) [inline aws/table_aws_emr_cluster.go:207]: Please follow the below order defining the schema
```
1. Add 4-5 columns at first which are important (i.e. name, arn, state etc.)
2. Add rest of the columns in alphabetical order,
3. Add all the JSON type columns at the end
4. For the standard column, follow this order (title, tags, akas)
```

File: aws/table_aws_emr_cluster.go [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+	op, err := svc.DescribeCluster(params)
+	if err != nil {
+		logger.Debug("getEmrCluster", "ERROR", err)
+		return nil, err
+	}
+
+	if op.Cluster != nil {
+		return op.Cluster, nil
+	}
```
Comment (Subhajit97) [inline aws/table_aws_emr_cluster.go:298]: Is this check required?
Doesn't the API return any `Not Found` errors?

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+data "null_data_source" "resource" {
+  inputs = {
+    scope = "arn:${data.aws_partition.current.partition}:::${data.aws_caller_identity.current.account_id}"
+  }
+}
+
+resource "aws_iam_role" "iam_emr_service_role" {
+  name               = "test_iam_emr_service_role"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:50]:
```suggestion
  name               = var.resource_name
```

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+        "Service": "elasticmapreduce.amazonaws.com"
+      },
+      "Action": "sts:AssumeRole"
+    }
+  ]
+}
+EOF
+}
+resource "aws_iam_role_policy" "iam_emr_service_policy" {
+  name = "test_iam_emr_service_policy"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:68]:
```suggestion
  name = var.resource_name
```

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+            "s3:CreateBucket",
+            "s3:Get*",
+            "s3:List*",
+            "sdb:BatchPutAttributes",
+            "sdb:Select",
+            "sqs:CreateQueue",
+            "sqs:Delete*",
+            "sqs:GetQueue*",
+            "sqs:PurgeQueue",
+            "sqs:ReceiveMessage"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:129]: Try to create with 2-3 read-only permissions

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+        ]
+    }]
+}
+EOF
+}
+resource "aws_iam_role" "iam_emr_profile_role" {
+  name = "test_iam_emr_profile_role"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:136]:
```suggestion
  name = var.resource_name
```

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+        "Service": "ec2.amazonaws.com"
+      },
+      "Action": "sts:AssumeRole"
+    }
+  ]
+}
+EOF
+}
+resource "aws_iam_instance_profile" "emr_profile" {
+  name  = "emr_profile_new008"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:154]:
```suggestion
  name  = var.resource_name
```

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+  role = "${aws_iam_role.iam_emr_profile_role.name}"
+}
+resource "aws_iam_role_policy" "iam_emr_profile_policy" {
+  name = "iam_emr_profile_policy"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:158]:
```suggestion
  name = var.resource_name
```

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+      },
+      "Action": "sts:AssumeRole"
+    }
+  ]
+}
+EOF
+}
+resource "aws_iam_instance_profile" "emr_profile" {
+  name  = "emr_profile_new008"
+  role = "${aws_iam_role.iam_emr_profile_role.name}"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:155]:
```suggestion
  role = aws_iam_role.iam_emr_profile_role.name
```

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+}
+resource "aws_iam_role_policy" "iam_emr_profile_policy" {
+  name = "iam_emr_profile_policy"
+  role = "${aws_iam_role.iam_emr_profile_role.id}"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:159]:
```suggestion
  role = "aws_iam_role.iam_emr_profile_role.id"
```

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+        ]
+    }]
+}
+EOF
+}
+resource "aws_vpc" "my_vpc" {
+  cidr_block           = "168.31.0.0/16"
+}
+resource "aws_security_group" "emr_master" {
+  vpc_id                 = "${aws_vpc.my_vpc.id}"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:198]:
```suggestion
  vpc_id                 = aws_vpc.my_vpc.id
```

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+  vpc_id = "${aws_vpc.my_vpc.id}"
+  route {
+    cidr_block = "0.0.0.0/0"
+    gateway_id = "${aws_internet_gateway.gw.id}"
+  }
+}
+resource "aws_main_route_table_association" "test_association" {
+  vpc_id         = "${aws_vpc.my_vpc.id}"
+  route_table_id = "${aws_route_table.test_route_table.id}"
+}
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:233]:
```suggestion
resource "aws_subnet" "my_subnet" {
  vpc_id     = aws_vpc.my_vpc.id
  cidr_block = "168.31.0.0/20"
}
resource "aws_internet_gateway" "gw" {
  vpc_id = aws_vpc.my_vpc.id
}
resource "aws_route_table" "test_route_table" {
  vpc_id = aws_vpc.my_vpc.id
  route {
    cidr_block = "0.0.0.0/0"
    gateway_id = aws_internet_gateway.gw.id
  }
}
resource "aws_main_route_table_association" "test_association" {
  vpc_id         = aws_vpc.my_vpc.id
  route_table_id = aws_route_table.test_route_table.id
}
```

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+resource "aws_emr_cluster" "named_test_resource" {
+  name          = var.resource_name
+  release_label = "emr-4.9.5"
+  applications  = ["Spark"]
+  ec2_attributes {
+    key_name = "${aws_key_pair.deployer.key_name}"
+    subnet_id                         = "${aws_subnet.my_subnet.id}"
+    emr_managed_master_security_group = "${aws_security_group.emr_master.id}"
+    emr_managed_slave_security_group  = "${aws_security_group.emr_master.id}"
+    instance_profile                  = "${aws_iam_instance_profile.emr_profile.arn}"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:248]:
```suggestion
    subnet_id                         = aws_subnet.my_subnet.id
    emr_managed_master_security_group = aws_security_group.emr_master.id
    emr_managed_slave_security_group  = aws_security_group.emr_master.id
    instance_profile                  = aws_iam_instance_profile.emr_profile.arn
```

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 20]
```diff
@@ -0,0 +1,289 @@
...
+    instance_type = "m4.xlarge"
+  }
+  core_instance_group {
+    instance_type  = "c4.large"
+    instance_count = 1
+  }
+  tags = {
+    name     = var.resource_name
+  }
+  service_role = "${aws_iam_role.iam_emr_service_role.arn}"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:260]:
```suggestion
  service_role = aws_iam_role.iam_emr_service_role.arn
```

File: aws-test/tests/aws_emr_cluster/variables.tf [hunk 21]
```diff
@@ -0,0 +1,289 @@
...
+  applications  = ["Spark"]
+  ec2_attributes {
+    key_name = "${aws_key_pair.deployer.key_name}"
+    subnet_id                         = "${aws_subnet.my_subnet.id}"
+    emr_managed_master_security_group = "${aws_security_group.emr_master.id}"
+    emr_managed_slave_security_group  = "${aws_security_group.emr_master.id}"
+    instance_profile                  = "${aws_iam_instance_profile.emr_profile.arn}"
+  }
+  master_instance_group {
+    instance_type = "m4.xlarge"
```
Comment (Subhajit97) [inline aws-test/tests/aws_emr_cluster/variables.tf:251]: Can we create resource of lower configuration?
```suggestion
    instance_type = "m1.small"
```

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
Make suggested changes

File: aws/table_aws_emr_cluster.go [hunk 22]
```diff
@@ -0,0 +1,330 @@
...
+
+	return op.Cluster, nil
+
+}
+
+//// TRANSFORM FUNCTIONS
+
+func getEmrClusterTurbotTags(ctx context.Context, d *transform.TransformData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("emrClusterTagToTurbotTags")
```
Comment (Subhajit97) [inline aws/table_aws_emr_cluster.go:303]:
```suggestion
	plugin.Logger(ctx).Trace("getEmrClusterTurbotTags")
```

File: docs/tables/aws_emr_cluster.md [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+
+### Basic info
+
+```sql
+select
+  id,
+  cluster_arn,
+  name,
+  auto_terminate,
+  status ->>  'State' as state,
```
Comment (Subhajit97) [inline docs/tables/aws_emr_cluster.md:15]:
```suggestion
  status ->> 'State' as state,
```

File: docs/tables/aws_emr_cluster.md [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+
+
+### List of EMR clusters which got terminated due to some error
+
+```sql
+select
+  id,
+  name,
+  status ->>  'State' as state,
+  status ->  'StateChangeReason' ->> 'Message'  as  state_change_reason
```
Comment (Subhajit97) [inline docs/tables/aws_emr_cluster.md:43]:
```suggestion
  status ->> 'State' as state,
  status ->> 'StateChangeReason' ->> 'Message'  as  state_change_reason
```

File: docs/tables/aws_emr_cluster.md [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+from
+  aws_emr_cluster
+where
+  status ->>  'State' = 'TERMINATED_WITH_ERRORS';
```
Comment (Subhajit97) [inline docs/tables/aws_emr_cluster.md:47]:
```suggestion
  status ->> 'State' = 'TERMINATED_WITH_ERRORS';
```

**Review by Subhajit97 — APPROVED** [review]
LGTM

File: aws/plugin.go [hunk 26]
```diff
@@ -67,6 +67,7 @@ func Plugin(ctx context.Context) *plugin.Plugin {
 			"aws_ec2_transit_gateway":                tableAwsEc2TransitGateway(ctx),
 			"aws_ec2_transit_gateway_route_table":    tableAwsEc2TransitGatewayRouteTable(ctx),
 			"aws_ec2_transit_gateway_vpc_attachment": tableAwsEc2TransitGatewayVpcAttachment(ctx),
+			"aws_emr_cluster":                        tableAwsEmrCluster(ctx),
```
Comment (Subhajit97) [inline aws/plugin.go:70]: Sort the list

**Review by Subhajit97 — DISMISSED** [review]
LGTM

**Review by Subhajit97 — APPROVED** [review]
LGTM

File: docs/tables/aws_emr_cluster.md [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+```sql
+select
+  name,
+  cluster_arn,
+  a ->> 'Name' as application_name,
+  a ->> 'Version' as application_version
+from
+  aws_emr_cluster,
+  jsonb_array_elements(applications) as a;
+```
```
Comment (rajmohanty17) [inline docs/tables/aws_emr_cluster.md:62]: Add one more example to check of EMR Cluster is Encrypted or not

Comment (rajmohanty17) [reply docs/tables/aws_emr_cluster.md:62]: @khushboo9024 pls check EMR Cluster Logging or not or by default.

**Review by rajmohanty17 — CHANGES_REQUESTED** [review]
Added comment to check encryption status.

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**